### PR TITLE
Update the Emogrifier dependency to ^2.0.0

### PR DESCRIPTION
**Pull request by oliverklee**
### Description
Emogrifer 2.0.0 does not introduce any backwards-compatibilty-breaking
API changes. This version adds new features, fixes bugs and improves
the resulting HTML. So the update should be reasonably safe.

### Manual testing scenarios
1. have Magento send an HTML email
2. see that the generated email still looks okay

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds on Travis CI are green)

(Should I as a contributor check those, or will the reviewers do this?)

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=13132) <br/>All committers have signed the CLA.

**Comment by oliverklee** [comment]
Just signed the CLA.

**Comment by oliverklee** [comment]
Okay, looks like I need to update the email generation error handling. Working on it.

**Comment by orlangur** [comment]
> Emogrifer 2.0.0 does not introduce any backwards-compatibilty-breaking

Is such policy applied for future versions too? I mean, is `^2.0.0` safe enough or maybe we should stick to `~2.0.0` instead.

> (Should I as a contributor check those, or will the reviewers do this?)

Yeah, feel free to check those, probably reviewer will uncheck some if he has another opinion.

**Comment by oliverklee** [comment]
> Is such policy applied for future versions too? I mean, is ^2.0.0 safe enough or maybe we should stick to ~2.0.0 instead.

Yes, `^2.0.0` will be safe enough. Emogrifier follows semantic versioning very strictly.

**Comment by orlangur** [comment]
@oliverklee,

> Emogrifier follows semantic versioning very strictly.

In such case we should make `~2.0.0` constraint and bump to `~2.1.0` and further manually with proper checks. So that we won't break Magento SemVer commitment and, for instance, receive breaking Emogrifier upgrade to 3.0 in a Magento patch release.

**Comment by oliverklee** [comment]
Rebased, updated the tests and added a setter for the Emogrifier debug mode.

**Comment by oliverklee** [comment]
> In such case we should make ~2.0.0 constraint and bump to ~2.1.0 and further manually with proper checks. So that we won't break Magento SemVer commitment and, for instance, receive breaking Emogrifier upgrade to 3.0 in a Magento patch release.

@orlangur I'm not sure I understand your concern correctly. Are you worried that `^2.0.0` would allow (breaking) updates to 3.0.0?

**Comment by orlangur** [comment]
@oliverklee yeah, it can happen with any `composer update` later.

**Comment by oliverklee** [comment]
@orlangur Thanks for the clarification. `^2.0.0` will only allow updates to 2.x.y, not to 3.x.y, and hence should be safe. Please see https://getcomposer.org/doc/articles/versions.md#caret-version-range- for the documentation of the `^` version constraint.

**Comment by orlangur** [comment]
@oliverklee will it allow Emogrifier upgrade to `2.1` in some of the `2.2.x` Magento patch releases?

**Comment by oliverklee** [comment]
@orlangur That's correct. And as Emogrifier follows semantic versioning, 2.1.x will not break anything over 2.0.0 (and will also not require any code changes in Magento).

I've noticed that some dependencies in the `composer.json` are set using `^`, while others are set using `~`. What's the general policy in the Magento project for using `^` vs. `~` for dependencies?

**Comment by oliverklee** [comment]
@orlangur And I'm also not sure whether my PR should go to the 2.2-develop branch, the 2.3-develop branch, or both. (I couldn't find anything about this in the CONTRIBUTING.md file.) Any advice on this? Thanks!

**Comment by orlangur** [comment]
@oliverklee just noticed there are many ZF component using caret constraint, let's leave the one proposed by you until we face with any issue.

> PR should go to the 2.2-develop branch, the 2.3-develop branch, or both

Any acceptable from SemVer perspective contribution must go to `2.2-develop` first, I'll check if this process is properly documented.

File: app/code/Magento/Email/Model/Template/Filter.php
Comment (orlangur) [inline app/code/Magento/Email/Model/Template/Filter.php]: Why do we need this?

**Comment by oliverklee** [comment]
@orlangur Okay, thanks for the update and the information!

**Review by orlangur — CHANGES_REQUESTED** [review]

File: app/code/Magento/Email/Model/Template/Filter.php
Comment (oliverklee) [reply app/code/Magento/Email/Model/Template/Filter.php]: Emogrifier 1.2 always threw an exception when it encountered something it could not parse (if it didn't directly bail with an XPath error). The filter then either caught and discarded the exception (in production mode) or added an error message (in non-production mode).

Emogrifier 2.0 now has a separate debug mode. In debug mode, it still throws exceptions. In the non-debug mode (the default), it silently ignores anything it cannot parse, and tries to cope the best it can.

So to have both old filter behaviors, we need to enable/disable the Emogrifier debug mode depending on the environment.

Comment (orlangur) [reply app/code/Magento/Email/Model/Template/Filter.php]: Got it, thanks for detailed explanation.

File: lib/internal/Magento/Framework/Css/PreProcessor/Adapter/CssInliner.php
Comment (orlangur) [inline lib/internal/Magento/Framework/Css/PreProcessor/Adapter/CssInliner.php]: Let's not expose too much methods when we don't really need it.

Please discard changes in `app/code/Magento/Email/Model/Template/Filter.php` and instead of introducing a new method here inject `AppState` and do `setDebug(true)` in constructor in case of developer mode.

**Review by orlangur — CHANGES_REQUESTED** [review]

File: lib/internal/Magento/Framework/Css/PreProcessor/Adapter/CssInliner.php
Comment (oliverklee) [reply lib/internal/Magento/Framework/Css/PreProcessor/Adapter/CssInliner.php]: On it.

Comment (oliverklee) [reply lib/internal/Magento/Framework/Css/PreProcessor/Adapter/CssInliner.php]: Done and repushed. I'm not very familiar with how Magento handles DI, so I'm keeping my fingers crossed.

**Comment by magento-engcom-team** [comment]
@oliverklee thank you for contributing. Please accept Community Contributors team invitation [here](https://github.com/orgs/magento/invitation) to gain extended permissions for this repository.

**Comment by oliverklee** [comment]
@orlangur After this has been merged, what's the process to get this in 2.3-develop, too? Should I create a PR for this?

**Comment by orlangur** [comment]
@oliverklee yeah, you can create a PR for `2.3-develop` even at the same time with `2.2-develop` one. Otherwise your changes will not be missed anyway, core team will port everything from `2.2-develop` to `2.3-develop` by themselves.

**Comment by oliverklee** [comment]
@orlangur Done: #13351

**Comment by oliverklee** [comment]
@orlangur I think this PR still has the review status "changes requested" (by you). If this is obsolete (which I believe), could you please update the review status? Thanks! :coffee:

**Review by orlangur — APPROVED** [review]

**Pull request merged**
